cupy/cuda/pinned_memory.pyx

```diff
@@ -0,0 +1,270 @@
+            function, which takes the buffer size as an argument and returns
+            the device buffer of that size.
+
+    """
+    global _current_allocator
+    _current_allocator = allocator
+
+
+cdef class PooledPinnedMemory(PinnedMemory):
+
+    """Memory allocation for a memory pool.
+
+    The instance of this class is created by memory pool allocator, so user
+    should not instantiate it by hand.
+
+    """
+
+    def __init__(self, PinnedMemory mem, pool):
+        self.ptr = mem.ptr
+        self.size = mem.size
+        self.pool = pool
+
+    def __dealloc__(self):
+        if self.ptr != 0:
+            self.free()
+
+    cpdef free(self):
+        """Frees the memory buffer and returns it to the memory pool.
+
+        This function actually does not free the buffer. It just returns the
+        buffer to the memory pool for reuse.
+
+        """
+        pool = self.pool()
+        if pool and self.ptr != 0:
+            pool.free(self.ptr, self.size)
+        self.ptr = 0
+        self.size = 0
+
+
+cdef class PinnedMemoryPool:
+
+    """Memory pool on the machine.
+
+    A memory pool preserves any allocations even if they are freed by the user.
+    Freed memory buffers are held by the memory pool as *free blocks*, and they
+    are reused for further memory allocations of the same sizes.
+
+    Args:
+        allocator (function): The base CuPy pinned memory allocator. It is
+            used for allocating new blocks when the blocks of the required
+            size are all in use.
+
+    """
+
+    def __init__(self, allocator=_malloc):
+        self._in_use = {}
+        self._free = collections.defaultdict(list)
+        self._alloc = allocator
+        self._weakref = weakref.ref(self)
+        self._allocation_unit_size = 512
+
+    cpdef PinnedMemoryPointer malloc(self, Py_ssize_t size):
+        cdef list free
+        cdef PinnedMemory mem
+
+        if size == 0:
+            return PinnedMemoryPointer(PinnedMemory(0), 0)
+
+        # Round up the memory size to fit memory alignment of cudaHostAlloc
+        unit = self._allocation_unit_size
+        size = (((size + unit - 1) // unit) * unit)
+        free = self._free[size]
+        if free:
+            mem = free.pop()
+        else:
+            try:
+                mem = self._alloc(size).mem
+            except runtime.CUDARuntimeError as e:
+                if e.status != 2:
+                    raise
+                self.free_all_free()
+                mem = self._alloc(size).mem
+
+        self._in_use[mem.ptr] = mem
+        pmem = PooledPinnedMemory(mem, self._weakref)
+        return PinnedMemoryPointer(pmem, 0)
+
+    cpdef free(self, size_t ptr, Py_ssize_t size):
+        cdef list free
+        cdef PinnedMemory mem
+        mem = self._in_use.pop(ptr, None)
+        if mem is None:
+            raise RuntimeError('Cannot free out-of-pool memory')
+        free = self._free[size]
+        free.append(mem)
+
+    cpdef free_all_free(self):
```

Comment:
``free_all_free`` is difficult to understand; how about replacing it with ``free_all_blocks``?